src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+            features.push("NOTIFY_ON_EMPTY");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ANY_LAYOUT) {
+            features.push("ANY_LAYOUT");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_INDIRECT_DESC) {
+            features.push("RING_INDIRECT_DESC");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_EVENT_IDX) {
+            features.push("RING_EVENT_IDX");
+        }
+        if sound_features.contains(SoundFeatures::UNUSED) {
+            features.push("UNUSED");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_VERSION_1) {
+            features.push("VERSION_1");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ACCESS_PLATFORM) {
+            features.push("ACCESS_PLATFORM");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_RING_PACKED) {
+            features.push("RING_PACKED");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_IN_ORDER) {
+            features.push("IN_ORDER");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ORDER_PLATFORM) {
+            features.push("ORDER_PLATFORM");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_SR_IOV) {
+            features.push("SR_IOV");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_NOTIFICATION_DATA) {
+            features.push("NOTIFICATION_DATA");
+        }
+        format!("[{}]", features.join(", "))
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct VirtIOSoundConfig {
+    jacks: ReadOnly<u32>,
+    streams: ReadOnly<u32>,
+    chmaps: ReadOnly<u32>,
+}
+
+/// In virtIO v1.2, this enum should be called "Code".
+///
+/// To avoid ambiguity in its meaning, I use the term "CommandCode" here.
+#[repr(u32)]
+#[derive(FromPrimitive, IntoPrimitive)]
+enum CommandCode {
+    /* jack control request types */
+    VirtioSndRJackInfo = 1,
+    VirtioSndRJackRemap,
+
+    /* PCM control request types */
+    VirtioSndRPcmInfo = 0x0100,
+    VirtioSndRPcmSetParams,
+    VirtioSndRPcmPrepare,
+    VirtioSndRPcmRelease,
+    VirtioSndRPcmStart,
+    VirtioSndRPcmStop,
+
+    /* channel map control request types */
+    VirtioSndRChmapInfo = 0x0200,
+
+    /* jack event types */
+    VirtioSndEvtJackConnected = 0x1000,
+    VirtioSndEvtJackDisconnected,
+
+    /* PCM event types */
+    VirtioSndEvtPcmPeriodElapsed = 0x1100,
+    VirtioSndEvtPcmXrun,
+
+    /* common status codes */
+    /// success
+    #[num_enum(default)]
+    VirtioSndSOk = 0x8000,
+    /// a control message is malformed or contains invalid parameters
+    VirtioSndSBadMsg,
+    /// requested operation or parameters are not supported
+    VirtioSndSNotSupp,
+    ///  an I/O error occurred
+    VirtioSndSIoErr,
+}
+
+/// Enum representing the types of item information requests.
+#[derive(Clone, Copy)]
+pub enum ItemInfomationRequestType {
```

Comment:
This looks like it should also be `#[repr(u32)]`. Also, you've made a typo in the name, it should be `ItemInformationRequestType`.

Reply:
Actually, how is this different to `CommandCode`? It seems like a subset of it.

Reply:
Yes, `ItemInformationRequestType` is indeed a subset of `CommandCode`. Unlike `CommandCode`, `ItemInformationRequestType` is public. There are some request codes in `CommandCode` that users do not need. Therefore, for user convenience, I selected its subset and defined it as a public enumeration.